### PR TITLE
Fix infinite asset preparation due to undrained AssetEvent events 

### LINKED ISSUE #11240: Texture in render_to_texture example doesn't update
## Bevy version

main (cfcb6885e3b475a93ec0fe7e88023ac0f354bbbf), since #10520

Seemingly not fixed by #11212

## Relevant system information

```
AdapterInfo { name: "Apple M1 Max", vendor: 0, device: 0, device_type: IntegratedGpu, driver: "", driver_info: "", backend: Metal }
SystemInfo { os: "MacOS 13.6.2 ", kernel: "22.6.0", cpu: "", core_count: "10", memory: "64.0 GiB" }
```

## What you did

`cargo run --example render_to_texture`

## What went wrong

Brown cube displayed on cube faces should also be doing its own rotation, but it's not.

### PR DESCRIPTION
**Pull request by brianreavis**
# Objective

After #10520, I was experiencing seriously degraded performance that ended up being due to never-drained `AssetEvent` events causing havoc inside `extract_render_asset::<A>`. The same events being read over and over again meant the same assets were being prepared every frame for eternity. For what it's worth, I was noticing this on a static scene about every 3rd or so time running my project.

* References #10520
* Fixes #11240

Why these events aren't sometimes drained between frames is beyond me and perhaps worthy of another investigation, but the approach in this PR effectively restores the original cached `EventReader` behavior (which fixes it).

## Solution

I followed the [`CachedSystemState` example](https://github.com/bevyengine/bevy/blob/3a666cab23eab5e83552ae074f47d8663221311e/crates/bevy_ecs/src/system/function_system.rs#L155) to make sure that the `EventReader` state is cached between frames like it used to be when it was an argument of `extract_render_asset::<A>`.

### CODE REVIEW

**Comment by mockersf** [comment]
this PR fixes https://github.com/bevyengine/bevy/issues/11240

**Review by torsteingrindvik — APPROVED** [review]
I checked the `render_to_texture` example on this PR and it is indeed fixed.

Also pointed my Bevy app which is broken on main towards this PR and that is fixed too :+1:

**Review by rodolphito — APPROVED** [review]
note: easier to review with whitespace diff hidden

File: crates/bevy_render/src/render_asset.rs [hunk 1]
```diff
@@ -176,49 +178,69 @@ impl<A: RenderAsset> RenderAssets<A> {
...
+struct CachedExtractRenderAssetSystemState<A: RenderAsset> {
+    state: SystemState<(
+        EventReader<'static, 'static, AssetEvent<A>>,
+        ResMut<'static, Assets<A>>,
+    )>,
+}
+
+impl<A: RenderAsset> FromWorld for CachedExtractRenderAssetSystemState<A> {
+    fn from_world(world: &mut bevy_ecs::world::World) -> Self {
+        Self {
```
Comment (JMS55) [inline crates/bevy_render/src/render_asset.rs:191]: Small nit: import World directly instead of writing out the full path

**Review by JMS55 — APPROVED** [review]

**Comment by Elabajaba** [comment]
I was seeing significant memory usage increases (VRAM and CPU), as well as significantly longer load times when testing loading large scenes on main, and this fixes them.

**Comment by alice-i-cecile** [comment]
Merging now to help y'all diagnose other perf issues more easily.

**Pull request merged**